### PR TITLE
Stablehlo to TF saved-model

### PR DESCRIPTION
**Pull request by sdasgup3**
Provides an API to convert stablehlo to tf saved model. Added a test to convert a simple linear model.

Things to me done:

- [x] Provide documentation of the API
- [x] Test should used the API from the preinstalled stablehlo python package
- [x] Make sure the API is provoded as part of stablehlo python wheel.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,7 +139,10 @@ If you'd like to build the Python bindings, you'll need to install a few
 additional dependencies.
 
 ```sh
-pip install  install -r ./llvm-project/mlir/python/requirements.txt
+pip install -r ./llvm-project/mlir/python/requirements.txt
+
+pip install pip install tensorflow-cpu # needed mainly for
+                                # stablehlo to tf-save-model connversion
```
Comment (GleasonK) [inline README.md:145]: This should be a separate section. Maybe even in a readme in the `savedmodel` directory.

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import mlir.dialects.stablehlo as stablehlo
+import mlir.ir as ir
+from mlir.stablehlo_to_tf_saved_model import InputLocation, save_stablehlo_as_tf_saved_model
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:17]: This doesn't make sense to me. If possible, I think we want it to be:

```
from stablehlo.savedmodel import stablehlo_to_tf_saved_model
```

But lets get it working first.

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+"""
+
+ctx = ir.Context()
+stablehlo.register_dialect(ctx)
+module = ir.Module.parse(mlir_module_string, ctx)
+
+input_locations = [
+    InputLocation.parameter(name='linear_layer.bias'),
+    InputLocation.parameter(name='linear_layer.weight'),
+    InputLocation.input_arg(position=0),
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:45]: Do we need to specify input shape anywhere?

Comment (sdasgup3) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:47]: The input shapes are derived from within the `save_stablehlo_as_tf_saved_model` using the module parameter of the API.

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py (hunk 2 above)
Comment (sdasgup3) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py]: I am not sure if we should put the `savedmodel` under `mlir.dialects`.
Currently the package is hoisted at `build/python_packages/stablehlo/savedmodel` which allows me to import it as:
 
```
from savedmodel.stablehlo_to_tf_saved_model import InputLocation, stablehlo_to_tf_saved_model
```

Please let me know what do you think?

File: README.md (hunk 1 above)
Comment (sdasgup3) [reply README.md]: Actually without this dependency the tests cannot be build when `STABLEHLO_ENABLE_BINDINGS_PYTHON` is ON. Added a clarification note in the curent readme file. PTAL.


Also, added a `README.md` at `stablehlo.integrations/python/savedmodel` to provide clarification on the dependency.

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import tempfile
+import mlir.dialects.stablehlo as stablehlo
+import mlir.ir as ir
+import numpy as np
+from savedmodel.stablehlo_to_tf_saved_model import InputLocation, stablehlo_to_tf_saved_model
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:20]: This still feels strange for `pip install stablehlo` to include a top-level `savedmodel` import? Can we move this to a `stablehlo` subdirectory at least?

File: README.md (hunk 1 above)
Comment (GleasonK) [reply README.md]: That's a merge-blocker IMO... we shouldn't (/ can't) require TF to build StableHLO.

Is it a build issue or test issue? We shouldn't require TF dependency for build. For testing we can define a separate test group (like `check-stablehlo-quick`) which requires the dependency.

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (sdasgup3) [reply README.md]: This is just a test issue. Created a separate section in the top level README about StableHLO TF integration and used stablehlo/savedmodel/README.md to explain the dependencies and testing instructions. PTAL.

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py (hunk 2 above)
Comment (sdasgup3) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py]: With the recent changes now the import looks like
```
from mlir.stablehlo.savedmodel.stablehlo_to_tf_saved_model import InputLocation, stablehlo_to_tf_saved_model
```
Please let me know your opinion.

Just to document:
Adding the nesting `stablehlo/savedmodel` parallel to mlir gives error like
```
ModuleNotFoundError: No module named 'stablehlo.savedmodel'; 'stablehlo' is not a package
```

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py (hunk 4 above)
Comment (sdasgup3) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py]: PTAL at https://github.com/openxla/stablehlo/pull/2157#discussion_r1556553712

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py (hunk 2 above)
Comment (GleasonK) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py]:
```
from mlir.stablehlo.savedmodel.stablehlo_to_tf_saved_model import InputLocation, stablehlo_to_tf_saved_model
```

LGTM!

Comment (sdasgup3) [reply stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py]: Yup this is what we have added now in the test file.

File: stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+  def __init__(self, spec: StableHLOFuncSpec):
+    self.stablehlo_type_to_tf_type = {
+        'i8': 'int8',
+        'i16': 'i32',
+        'i32': 'int32',
+        'i64': 'int64',
+        'f16': 'float16',
+        'f32': 'float32',
+        'f64': 'float64',
```
Comment (GleasonK) [inline stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py:91]: Any missing types?

bf16, f16, fp8 types ?

File: stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+  stablehlo_type_to_tf_type = {
+      'i8': 'int8',
+      'i16': 'i32',
+      'i32': 'int32',
+      'i64': 'int64',
+      'f16': 'float16',
+      'f32': 'float32',
+      'f64': 'float64',
+  }
```
Comment (GleasonK) [inline stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py:111]: Redefinition of above?

File: stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+      if str(loc.type_) == str(VariableType.PARAMETER):
+        call_args.append(self.stablehlo_program.state_dict[loc.name])
+      else:
+        call_args.append(args[loc.position])
+    return call_args
+
+  def _wrap_as_tf_func(self):
+    def inner(*args):
+      Touts = [
+          self.stablehlo_type_to_tf_type[sig.dtype]
```
Comment (GleasonK) [inline stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py:125]: Do we crash on unsupported types? Should we add error checking?

File: stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+# Top level API for stablehlo to tf saved model
+
+
+def stablehlo_to_tf_saved_model(
+    module: ir.Module,
+    saved_model_dir: os.PathLike,
+    input_locations: list = [],
+    state_dict: dict = {},
+):
+  target = stablehlo.get_current_version()
```
Comment (GleasonK) [inline stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py:216]: I think TF has some forward compat.. 1month at least. We could probably make this an input argument that defaults to current

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import mlir.ir as ir
+import numpy as np
+from mlir.stablehlo.savedmodel.stablehlo_to_tf_saved_model import InputLocation, stablehlo_to_tf_saved_model
+import tensorflow as tf
+from tensorflow.python.tools import saved_model_utils
+
+# Convert a stablehlo program, expressing a nn.Linear layer with constant values
+# for weight and bias, to saved model.
+
+mlir_module_string = """
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:27]: `MODULE_STRING = `

We tend to use caps for these sort of constants

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+stablehlo_to_tf_saved_model(
+    module,
+    saved_model_dir=saved_model_dir,
+    input_locations=input_locations,
+    state_dict=state_dict,
+)
+
+saved_model = saved_model_utils.read_saved_model(saved_model_dir)
+assert saved_model != None
+print(f'StableHLO convertion to TF Saved Model seems to work!')
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:67]: How about load savedmodel and call? Why don't we swap out the linear FP model for a more basic `add` model?

Example on load and call:
https://openxla.org/stablehlo/tutorials/jax-export#reload_and_call_the_savedmodel

**Review by GleasonK — CHANGES_REQUESTED** [review]
LGTM a few minor changes.

File: .github/workflows/buildAndTestCMake.yml [hunk 11]
```diff
@@ -93,6 +93,7 @@ jobs:
     - name: Build and Test StableHLO (with Python bindings)
       shell: bash
       run: |
+        pip install tensorflow-cpu
         ./build_tools/github_actions/ci_build_cmake.sh "$LLVM_BUILD_DIR" "$STABLEHLO_BUILD_DIR"
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply .github/workflows/buildAndTestCMake.yml:97]: How about we hide the installation of `pip install tensorflow-cpu` inside `build_tools/github_actions/ci_build_cmake.sh` once STABLEHLO_ENABLE_TF is true?


In any case, I have updated the README.md to reflect that

```
## Python package dependencies
...

* TensorFlow: Only used to work with TF saved model artifacts.

```

File: stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    module,
+    saved_model_dir=saved_model_dir,
+    input_locations=input_locations,
+    state_dict=state_dict,
+)
+
+restored_model = tf.saved_model.load(saved_model_dir)
+restored_result = restored_model.f(tf.constant([3], tf.float32))
+assert np.allclose(restored_result[0], tf.constant([6], tf.float32))
+print(f'StableHLO convertion to TF Saved Model seems to work!')
```
Comment (GleasonK) [inline stablehlo/integrations/python/tests/stablehlo_to_tf_saved_model_test.py:64]: Delete this print?

**Review by GleasonK — APPROVED** [review]

File: .github/workflows/buildAndTestCMake.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (GleasonK) [reply .github/workflows/buildAndTestCMake.yml:97]: No need for that. Lets leave it as-is, we can adjust if this causes any issues.

**Pull request merged**